README.md
# plugin.video.tv4.se
XBMC/Kodi addon for tv4play.se.

## Installation

Install the addon by downloading the addon as a zip file. 

In Github: Choose Clone or download->Download zip and save to a location of choice.

In Kodi: Goto System->Settings->Add-ons->Install from zip file. Select the downloaded zip file.


changelog.txt
28.6.2016 Initial version with basic support for streaming videos from tv4play.se


playapi.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
#    Tomas Melin 2016
#    Basic tv4play API implementation.
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this Program; see the file LICENSE.txt.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
from datetime import datetime
import time
import re
import xml.etree.ElementTree as ET
from urlparse import urlparse
import json
import os
import sys
import urllib
import urllib2
try:
    import xbmc
except:
    class xbmc(object):
        @staticmethod
        def log(message):
            sys.stdout.write(message)
            sys.stdout.write('\n')

class tv4PlayApi():
    USER_AGENT = 'Mozilla/5.0(iPad; U; CPU iPhone OS 3_2 like Mac OS X; en-us) ' \
                 'AppleWebKit/531.21.10 (KHTML, like Gecko) Version/4.0.4 Mobile/7B314 Safari/531.21.10'
    PROGRAMS = 'http://webapi.tv4play.se/play/programs?is_active=true&per_page=550&fl=name,nid,program_image,description,category&start=0'
    EPISODES_BASEURL = 'http://webapi.tv4play.se/play/video_assets?sort_order=desc&start=0&'
    VIDEO_BASEURL = 'http://prima.tv4play.se/api/web/asset/{0}/play'

    def __init__(self):
        xbmc.log('Starting tv4PlayApi')

    def get_jsondata(self, url):

        content = self.http_request(url)
        if content is not None and content != '':
            try:
                data = json.loads(content)
                return data
            except Exception as ex:
                raise tv4PlayApiException(ex)
        else:
            return []

    def http_request(self, url):
        xbmc.log('Requesting URL: {0}'.format(url))
        request = urllib2.Request(url, headers={
            'user-agent': tv4PlayApi.USER_AGENT,
            'Content-Type': 'application/vnd.api+json'
        })
        connection = urllib2.urlopen(request)
        content = connection.read()
        connection.close()
        return content

    def get_programs(self):
        programs = self.get_jsondata(tv4PlayApi.PROGRAMS)
        if programs['total_hits'] == '0':
            return []

        return programs['results']

    def get_episodes(self, program):
        program_url = tv4PlayApi.EPISODES_BASEURL + urllib.urlencode({'type': 'episode','q': program})
        data = self.get_jsondata(program_url)

        return data['results']

    def get_videodata(self, vid):
        url = tv4PlayApi.VIDEO_BASEURL.format(vid)
        try:
            data = self.http_request(url)
        except urllib2.HTTPError, e:
            data = e.read()
            xml = ET.XML(data)
            code = xml.find('code').text
            raise tv4PlayApiException(code)

        xml = ET.XML(data)
        ss = xml.find('items')
        sa = list(ss.iter('item'))

        if xml.find('drmProtected').text == 'true':
            msg = 'DRM_PROTECTED'
            raise tv4PlayApiException(msg)
        if xml.find('playbackStatus').text == 'NOT_STARTED':
            msg = 'PLAYBACKSTATUS'
            broadcast_time = xml.find('liveBroadcastTime').text
            format = '%Y-%m-%dT%H:%M:%S+02:00'
            #Workarond for strange TypeError
            #These are equivalent according to https://docs.python.org/2/library/datetime.html
            try:
                broadcast_time = datetime.strptime(broadcast_time, format).strftime('%c')
            except TypeError:
                broadcast_time = datetime(*(time.strptime(broadcast_time, format)[0:6])).strftime('%c')

            self.live_broadcast_time = broadcast_time
            raise tv4PlayApiException(msg)

        if xml.find('live').text == 'true':
            self.video_format = 'livehls'
        else:
            self.video_format = 'mp4'

        urls = self.get_videourls(vid)
        return urls

    def get_videourls(self, vid):

        url = tv4PlayApi.VIDEO_BASEURL.format(vid) + '?' + urllib.urlencode({'protocol': 'hls'})
        data = self.http_request(url)

        videourl = self.get_media_format_url(data, self.video_format)
        if videourl == '':
            msg = "NO_URL_FOUND"
            raise tv4PlayApiException(msg)

        subtitleurl = self.get_media_format_url(data, 'smi')

        return {'videourl' : videourl, 'subtitleurl' : subtitleurl}

    def get_media_format_url(self, data, field):
        xml = ET.XML(data)
        ss = xml.find('items')
        sa = list(ss.iter('item'))
        for i in sa:
            if i.find('mediaFormat').text == field:
                url = i.find('url').text
                return url

        return ''

    def get_start_time(self):
        return self.live_broadcast_time

class tv4PlayApiException(Exception):
    pass

if __name__ == '__main__':
    sys.exit(0)


addon.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# Based on code by:
#      Tommy Winther
#      Anders Norman
#
#  (c) Tomas Melin 2016
#
#  This Program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2, or (at your option)
#  any later version.
#
#  This Program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this Program; see the file LICENSE.txt.  If not, write to
#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#  http://www.gnu.org/copyleft/gpl.html
#
import os
import sys
import urlparse
import re
import urllib

import xbmc
import xbmcgui
import xbmcaddon
import xbmcplugin

from playapi import tv4PlayApi, tv4PlayApiException

class TV4PlayAddon():
    def __init__(self):
        self.api = tv4PlayApi()

    def _build_url(self, query):
        return PATH + '?' + urllib.urlencode(query)

    def list_programs(self):
        xbmcplugin.setContent(HANDLE, 'movies')
        programs = self.api.get_programs()
        if not programs:
            xbmcplugin.endOfDirectory(HANDLE, succeeded=False)
            self.display_error(30000)
            return

        items = list()
        for program in programs:
            fanart = ''
            if 'program_image' in program:
                fanart = program['program_image']
            infoLabels = {
                'title': program['name'],
                'plot' : program['description'],
                'genre' : program['category']['name']
            }
            item = xbmcgui.ListItem(program['name'], iconImage=fanart)
            item.setProperty('Fanart_Image', fanart)
            item.setInfo('video', infoLabels)
            url = self._build_url({'episodes_nid': program['nid'].encode('utf-8')})
            items.append((url, item, True))

        xbmcplugin.addDirectoryItems(HANDLE, items)
        xbmcplugin.addSortMethod(HANDLE, xbmcplugin.SORT_METHOD_TITLE)
        xbmcplugin.endOfDirectory(HANDLE)

    def list_episodes(self, program_nid):
        xbmcplugin.setContent(HANDLE, 'episodes')
        items = []
        episodes = self.api.get_episodes(program_nid)

        for episode in episodes:
            fanart = episode['image']

            info_labels = {
                'title': episode['title'],
                'plot' : episode['description'],
                'aired' : episode['broadcast_date_time'],
                'genre' : episode['program']['category']['name']
            }

            item = xbmcgui.ListItem(episode['title'], iconImage=fanart)
            item.setInfo('video', info_labels)
            item.setProperty('Fanart_Image', fanart)
            url = self._build_url({'play_video': episode['id']})
            item.setProperty('IsPlayable', 'true')
            items.append((url, item))

        xbmcplugin.addSortMethod(HANDLE, xbmcplugin.SORT_METHOD_EPISODE)
        xbmcplugin.addSortMethod(HANDLE, xbmcplugin.SORT_METHOD_TITLE)
        xbmcplugin.addSortMethod(HANDLE, xbmcplugin.SORT_METHOD_DATE)
        xbmcplugin.addDirectoryItems(HANDLE, items)
        xbmcplugin.endOfDirectory(HANDLE)

    def display_error(self, message='N/A', extra_info=''):
        heading = ADDON.getLocalizedString(30001)
        line1 = ADDON.getLocalizedString(30003)
        if str(message).isdigit():
           message = ADDON.getLocalizedString(int(message))
        xbmcgui.Dialog().ok(heading, line1, unicode(message), extra_info)

    def play_video(self, url):
        try:
            videodata = self.api.get_videodata(url)
            player = xbmc.Player()
            item = xbmcgui.ListItem('Video', path=videodata['videourl'])
            xbmcplugin.setResolvedUrl(HANDLE, True, item)

            if videodata['subtitleurl'] != "":
                xbmc.log('setting_subtitles: {0}'.format(videodata['subtitleurl']))
                player.setSubtitles(videodata['subtitleurl'])

                xbmc.log('subtitle-stream: {0}'.format(player.getAvailableSubtitleStreams()))

        except tv4PlayApiException as code:
            if 'SESSION_NOT_AUTHENTICATED' in code:
                self.display_error(30004)
            elif 'ASSET_PLAYBACK_INVALID_GEO_LOCATION' in code:
                self.display_error(30005)
            elif 'DRM_PROTECTED' in code:
                self.display_error(30007)
            elif 'PLAYBACKSTATUS' in code:
                self.display_error(30008, self.api.get_start_time())
            elif 'NO_URL_FOUND' in code:
                self.display_error(30009)
            else:
                self.display_error('{0}'.format(code))

if __name__ == '__main__':
    ADDON = xbmcaddon.Addon()
    PATH = sys.argv[0]
    HANDLE = int(sys.argv[1])
    PARAMS = urlparse.parse_qs(sys.argv[2][1:])

    ICON = os.path.join(ADDON.getAddonInfo('path'), 'icon.png')
    FANART = os.path.join(ADDON.getAddonInfo('path'), 'fanart.jpg')

    CACHE_PATH = xbmc.translatePath(ADDON.getAddonInfo('Profile'))
    if not os.path.exists(CACHE_PATH):
        os.makedirs(CACHE_PATH)

    tv4playAddon = TV4PlayAddon()
    try:
        if 'play_video' in PARAMS:
            tv4playAddon.play_video(PARAMS['play_video'][0])
        elif 'episodes_nid' in PARAMS:
            tv4playAddon.list_episodes(PARAMS['episodes_nid'][0])
        else:
            tv4playAddon.list_programs()

    except Exception as ex:
        tv4playAddon.display_error(str(ex))
